File: src/ai/flows/analyze-contract-risk.ts
'use server';
/**
 * @fileOverview Analyzes contract documents and identifies potential risks, key dates, parties, and definitions.
 *
 * - analyzeContractRisk - A function that analyzes contract risk.
 * - AnalyzeContractRiskInput - The input type for the analyzeContractRisk function.
 * - AnalyzeContractRiskOutput - The return type for the analyzeContractRisk function.
 */

import {ai} from '@/ai/ai-instance';
import {z} from 'genkit';
import {getDocumentMetadata} from '@/services/document-metadata';

const AnalyzeContractRiskInputSchema = z.object({
  documentText: z.string().describe('The text content of the contract document.'),
});
export type AnalyzeContractRiskInput = z.infer<typeof AnalyzeContractRiskInputSchema>;

const AnalyzeContractRiskOutputSchema = z.object({
  overallRiskScore: z.number().describe('The overall risk score of the contract (0-100).'),
  riskClauses: z.array(
    z.object({
      clauseText: z.string().describe('The text of the risky clause.'),
      riskCategory: z.string().describe('The category of risk associated with the clause.'),
      confidenceScore: z.number().describe('The confidence score (0-1) of the risk assessment.'),
      remediationSuggestion: z.string().describe('A suggested redline to mitigate the risk.'),
      clarifyingQuestion: z.string().describe('A clarifying question to ask about the clause.'),
    })
  ).describe('A list of clauses identified as potentially risky.'),
  keyDates: z.array(z.string().describe('Key dates identified in the contract related to risk.')).describe('Key dates related to risk'),
  parties: z.array(z.string().describe('Parties identified in the contract.')).describe('Parties involved in the contract.'),
  definitions: z.record(z.string(), z.string()).describe('A dictionary of definitions found in the contract.'),
  effectiveDate: z.string().optional().describe('The effective date of the contract, if found.'),
  executiveSummary: z.object({
    overall: z.string().describe('Overall plain-English summary of the contract risk.'),
    sections: z.array(z.object({
      sectionTitle: z.string(),
      summary: z.string()
    })).describe('Plain-English summaries per section of the contract.')
  }).describe('Executive summaries of the contract.'),
  regulatoryComplianceScore: z.number().describe('A score (0-100) indicating the contract\'s compliance with relevant regulations.'),
});
export type AnalyzeContractRiskOutput = z.infer<typeof AnalyzeContractRiskOutputSchema>;

export async function analyzeContractRisk(input: AnalyzeContractRiskInput): Promise<AnalyzeContractRiskOutput> {
  return analyzeContractRiskFlow(input);
}

const extractContractMetadata = ai.defineTool({
  name: 'extractContractMetadata',
  description: 'Extracts key metadata from a contract document, including key dates, parties involved, and definitions.',
  inputSchema: z.object({
    contractText: z.string().describe('The text of the contract to analyze.'),
  }),
  outputSchema: z.object({
    keyDates: z.array(z.string().describe('Key dates identified in the contract.')),
    parties: z.array(z.string().describe('Parties identified in the contract.')),
    definitions: z.record(z.string(), z.string()).describe('A dictionary of definitions found in the contract.'),
    effectiveDate: z.string().optional().describe('The effective date of the contract, if found.'),
  }),
}, async input => {
  const metadata = await getDocumentMetadata(input.contractText);
  return {
    keyDates: metadata.detectedDates,
    parties: metadata.parties,
    definitions: { /* TODO: Implement extraction of definitions */ },
    effectiveDate: metadata.effectiveDate,
  };
});

const generateExecutiveSummary = ai.defineTool({
  name: 'generateExecutiveSummary',
  description: 'Generates an executive summary of the contract, both overall and per section.',
  inputSchema: z.object({
    contractText: z.string().describe('The text of the contract to analyze.'),
  }),
  outputSchema: z.object({
    overall: z.string().describe('Overall plain-English summary of the contract risk.'),
    sections: z.array(z.object({
      sectionTitle: z.string(),
      summary: z.string()
    })).describe('Plain-English summaries per section of the contract.')
  }),
}, async input => {
  // TODO: Implement the logic to generate executive summaries.
  return {
    overall: 'Overall risk is moderate. Key areas to review are...',
    sections: [{ sectionTitle: 'Payment Terms', summary: 'Payment terms are standard.' }]
  };
});

const assessRegulatoryCompliance = ai.defineTool({
  name: 'assessRegulatoryCompliance',
  description: 'Assesses the contract\'s compliance with relevant regulations.',
  inputSchema: z.object({
    contractText: z.string().describe('The text of the contract to analyze.'),
  }),
  outputSchema: z.number().describe('A score (0-100) indicating the contract\'s compliance with relevant regulations.'),
}, async input => {
  // TODO: Implement the logic to assess regulatory compliance.
  return 75; // Example compliance score
});

const prompt = ai.definePrompt({
  name: 'analyzeContractRiskPrompt',
  tools: [extractContractMetadata, generateExecutiveSummary, assessRegulatoryCompliance],
  input: {
    schema: z.object({
      documentText: z.string().describe('The text content of the contract document.'),
    }),
  },
  output: {
    schema: AnalyzeContractRiskOutputSchema,
  },
  prompt: `You are an expert legal analyst specializing in identifying risks within contracts.

  Analyze the following contract text and identify potential risks. Provide an overall risk score (0-100) and highlight specific clauses that are considered risky, along with their risk categories and confidence scores. For each risky clause, generate a recommended redline to mitigate the risk and a clarifying question to ask about the clause.

  Contract Text: {{{documentText}}}

  Here are some risk patterns to look for: auto-renewal, force majeure, and security obligations.

  Use the extractContractMetadata tool to identify key dates, parties involved, and definitions within the contract.
  Use the generateExecutiveSummary tool to create an overall and per-section plain-English summary.
  Use the assessRegulatoryCompliance tool to get a regulatory compliance score.

  Include all extracted metadata and summaries in the output.

  Format your output as a JSON object matching the schema above.`,
});

const analyzeContractRiskFlow = ai.defineFlow<
  typeof AnalyzeContractRiskInputSchema,
  typeof AnalyzeContractRiskOutputSchema
>({
  name: 'analyzeContractRiskFlow',
  inputSchema: AnalyzeContractRiskInputSchema,
  outputSchema: AnalyzeContractRiskOutputSchema,
}, async input => {
  const {output} = await prompt(input);
  return {
    ...output!,
    riskClauses: output!.riskClauses.map(clause => ({
      ...clause,
      remediationSuggestion: 'Consider narrowing the scope of this clause...', // Example suggestion
      clarifyingQuestion: 'What specific scenarios does this clause intend to cover?' // Example question
    }))
  };
});

File: services/analyze_contract_risk.py
from typing import List, Dict, Optional, Any, Union
from pydantic import BaseModel, Field

class AnalyzeContractRiskInput(BaseModel):
    document_text: str = Field(..., description="The text content of the contract document.")
    contract_type: str = Field("NDA", description="The type of contract being analyzed.")

class RiskItem(BaseModel):
    clause_text: str = Field(..., description="The text of the risky clause.")
    risk_category: str = Field(..., description="The category of risk associated with the clause.")
    confidence_score: float = Field(..., description="The confidence score (0-1) of the risk assessment.")
    suggested_fix: str = Field(..., description="A suggested redline to mitigate the risk.")
    clarifying_question: str = Field(..., description="A clarifying question to ask about the clause.")
    industry_prevalence: str = Field(..., description="Comparison to industry prevalence (e.g., 'More common than 80% of similar MSAs').")
    priority: Optional[int] = Field(None, description="Priority of the risk item (1-3, lower is higher priority).")

class SectionSummary(BaseModel):
    section_title: str
    summary: str
    highest_risk_clause: Optional[str] = None

class ExecutiveSummary(BaseModel):
    overall: str = Field(..., description="Overall plain-English summary of the contract risk.")
    sections: List[SectionSummary] = Field(..., description="Plain-English summaries per section of the contract.")

class Metadata(BaseModel):
    effective_date: Optional[str] = None
    renewal_deadline: Optional[str] = None
    opt_out_deadline: Optional[str] = None
    parties: List[str] = Field([], description="Parties involved in the contract.")
    governing_law: Optional[str] = None
    venue: Optional[str] = None
    definitions: Dict[str, str] = Field({}, description="A dictionary of definitions found in the contract.")
    sla_references: List[str] = Field([], description="SLA references, if any.")

class AnalyzeContractRiskOutput(BaseModel):
    metadata: Metadata = Field(..., description="Comprehensive metadata of the contract.")
    executive_summary: ExecutiveSummary = Field(..., description="Executive summaries of the contract.")
    risk_items: List[RiskItem] = Field(..., description="A list of clauses identified as potentially risky, with remediation and questions.")
    overall_risk_score: float = Field(..., description="The overall risk score of the contract (0-100).")
    regulatory_compliance_score: float = Field(..., description="A score (0-100) indicating the contract's compliance with relevant regulations.")
    negotiation_points: List[str] = Field(..., description="Key points for negotiation.")

async def analyzeContractRisk(input_data: AnalyzeContractRiskInput) -> AnalyzeContractRiskOutput:
    """
    Analyzes contract text for risk factors and provides recommendations.
    
    Args:
        input_data: Contract text and analysis parameters
        
    Returns:
        Risk analysis results including scores and recommendations
    """
    # This is a placeholder implementation
    # In a real application, you would integrate with actual analysis services
    
    # Example metadata
    metadata = Metadata(
        effective_date="2025-01-01",
        renewal_deadline="2025-12-31",
        parties=["Acme Corporation", "Widget Inc."],
        governing_law="Delaware",
        venue="Wilmington",
        definitions={"Confidential Information": "All non-public information disclosed by either party."}
    )
    
    # Example executive summary
    executive_summary = ExecutiveSummary(
        overall="This contract contains standard confidentiality terms but lacks key provisions.",
        sections=[
            SectionSummary(
                section_title="PURPOSE",
                summary="The purpose is to explore a potential business relationship.",
                highest_risk_clause=None
            ),
            SectionSummary(
                section_title="DEFINITION",
                summary="Defines Confidential Information broadly.",
                highest_risk_clause="Broad scope of confidentiality"
            ),
            SectionSummary(
                section_title="OBLIGATIONS",
                summary="Outlines the obligations of each party.",
                highest_risk_clause="Obligations Standard"
            )
        ]
    )
    
    # Example risk items (similar to the TypeScript implementation)
    risk_items = []
    contract_text = input_data.document_text.lower()
    
    # 1. Missing Injunctive Relief (Priority 1)
    if "injunctive relief" not in contract_text and "equitable remedies" not in contract_text:
        risk_items.append(RiskItem(
            clause_text="Absence of Injunctive Relief Clause",
            risk_category="Enforcement",
            confidence_score=0.95,
            suggested_fix="Include a clause allowing for injunctive relief in the event of a breach.",
            clarifying_question="What remedies are available to protect confidential information?",
            industry_prevalence="Less common than 20% of similar NDAs",
            priority=1
        ))
    
    # 2. Data Return/Destroy Timelines (Priority 2)
    if "return or destroy" not in contract_text and "data retention" not in contract_text:
        risk_items.append(RiskItem(
            clause_text="Absence of Data Return/Destroy Clause",
            risk_category="Data Handling",
            confidence_score=0.92,
            suggested_fix="Include specific timelines for returning or destroying confidential information.",
            clarifying_question="What procedures are in place for the return of confidential information?",
            industry_prevalence="Less common than 30% of similar NDAs",
            priority=2
        ))
    
    # Calculate overall risk score
    overall_risk_score = 75.0  # Default score for illustration
    regulatory_compliance_score = 95.0  # Default score for illustration
    
    # Example negotiation points
    negotiation_points = [
        "Add injunctive relief / equitable remedies",
        "Ensure data-return/destroy timelines are clear and reasonable",
        "Removal of unilateral fees"
    ]
    
    return AnalyzeContractRiskOutput(
        metadata=metadata,
        executive_summary=executive_summary,
        risk_items=risk_items,
        overall_risk_score=overall_risk_score,
        regulatory_compliance_score=regulatory_compliance_score,
        negotiation_points=negotiation_points
    )